networkit/cpp/matching/BSuitorMatcher.cpp

```diff
@@ -0,0 +1,145 @@
+    if (G.numberOfSelfLoops() > 0)
+        throw std::runtime_error("This algorithm does not support graphs with self-loops.");
+
+    if (G.isDirected())
+        throw std::runtime_error("This algorithm does not support directed graphs.");
+
+    if (b.size() != G.numberOfNodes())
+        throw std::runtime_error(
+            "The number of b values does not match the number of nodes in this graph.");
+
+    const auto n = G.upperNodeIdBound();
+    Suitors.reserve(n);
+    Proposed.reserve(n);
+    for (index i = 0; i < n; i++) {
+        Suitors.emplace_back(std::make_unique<NodeMatchesInfo>(b.at(i)));
+        Proposed.emplace_back(std::make_unique<NodeMatchesInfo>(b.at(i)));
+    }
+}
+
+BSuitorMatcher::BSuitorMatcher(const Graph &G, count b)
+    : BSuitorMatcher(G, std::vector<count>(G.numberOfNodes(), b)) {}
+
+BSuitorMatcher::BSuitorMatcher(const Graph &G, const std::string &path)
+    : BSuitorMatcher(G, readBValuesFromFile(G.numberOfNodes(), path)) {}
+
+std::vector<count> BSuitorMatcher::readBValuesFromFile(count size, const std::string &path) const {
+    std::vector<count> b;
+    b.reserve(size);
+    std::ifstream file(path);
+    std::string line;
+    int line_number = 1;
+
+    while (std::getline(file, line)) {
+        std::istringstream istring(line);
+        int val;
+        if (!(istring >> val)) {
+            throw std::runtime_error("File " + path + " contains an invalid value in line "
+                                     + std::to_string(line_number) + ".");
+        }
+        if (istring >> val) {
+            throw std::runtime_error("File " + path + " contains multiple values in line "
+                                     + std::to_string(line_number) + ".");
+        }
+        if (val < 0) {
+            throw std::runtime_error("File " + path + " contains a negative value in line "
+                                     + std::to_string(line_number) + ".");
+        }
+        b.emplace_back(val);
+        line_number++;
+    }
+    if (b.size() != size) {
+        throw std::runtime_error("The number of values in file " + path
+                                 + " does not match the number of nodes in this graph.");
+    }
+    return b;
+}
+
+void BSuitorMatcher::findSuitors(node cur) {
+    for (index i = 0; i < b.at(cur); i++) {
+        auto [pref, heaviest] = findPreferred(cur);
+        if (pref != none) {
+            makeSuitor(cur, heaviest, pref);
+        }
+    }
+}
+
+Node BSuitorMatcher::findPreferred(node u) {
+    Node best = Node{none, 0};
+
+    auto hasProposedTo = [&](node x) -> bool {
+        return std::any_of(Proposed.at(u)->partners.begin(), Proposed.at(u)->partners.end(),
+                           [x](const Node &y) { return y.id == x; });
+    };
+
+    for (auto n : G->weightNeighborRange(u)) {
+        const Node v = Node(n.first, n.second);
+        if (!hasProposedTo(v.id)) {
```

Comment:
You can save an indentation level with:
```suggestion
        if (hasProposedTo(v.id)) continue;
```